### ISSUE TITLE
[JDBC] - storeItemValueProvider does not store Time for DATETIMEITEM 

### ISSUE THREAD

**Issue by sciamachy** [openhab/openhab1-addons#5255]
Ubuntu 17.04, java-8-openjdk-amd64, Openhab 2.1, JDBC Persistence 1.10.

When persisting a DATETIMEITEM, only the date (yyyy-mm-dd) is set.  It seems that in JdbcBaseDAO, storeItemValueProvider(), that the java.sql.Date does not return time.  Modifying to store javaType java.sql.Timestamp seems to resolve this issue:

        } else if ("DATETIMEITEM".equals(itemType)) {
            // vo.setValueTypes(getSqlTypes().get(itemType), java.util.Date.class);
            vo.setValueTypes(getSqlTypes().get(itemType), java.sql.Timestamp.class);
            Calendar x = ((DateTimeType) item.getState()).getCalendar();
            java.sql.Timestamp d = new java.sql.Timestamp(x.getTimeInMillis());
            logger.debug("JDBC::storeItemValueProvider: DateTimeItem: '{}'", d);
            vo.setValue(d);

Debug Log of Issue:
2017-07-17 21:36:56.681 [DEBUG] [persistence.jdbc.internal.JdbcMapper] - JDBC::storeItemValue: item=GarageMandoorLHF (Type=DateTimeItem, State=2017-07-17T21:36:56)
2017-07-17 21:36:56.682 [DEBUG] [penhab.persistence.jdbc.model.ItemVO] - JDBC:ItemVO tableName=item0041; newTableName=null;
2017-07-17 21:36:56.682 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::getItemType: Try to use ItemType DATETIMEITEM for Item GarageMandoorLHF
2017-07-17 21:36:56.682 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::storeItemValueProvider: item 'GarageMandoorLHF' as Type 'DATETIMEITEM' in 'item0041' with state '2017-07-17T21:36:56'
2017-07-17 21:36:56.683 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::storeItemValueProvider: getState: '2017-07-17T21:36:56'
2017-07-17 21:36:56.683 [DEBUG] [penhab.persistence.jdbc.model.ItemVO] - JDBC:ItemVO setValueTypes dbType=TIMESTAMP(3); javaType=class java.sql.Date;
2017-07-17 21:36:56.684 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::storeItemValueProvider: DateTimeItem: '2017-07-17'
2017-07-17 21:36:56.684 [DEBUG] [nhab.persistence.jdbc.db.JdbcBaseDAO] - JDBC::doStoreItemValue sql=INSERT INTO item0041 (TIME, VALUE) VALUES( NOW(3), ? ) ON DUPLICATE KEY UPDATE VALUE= ? value='2017-07-17'
2017-07-17 21:36:56.686 [DEBUG] [jdbc.internal.JdbcPersistenceService] - JDBC: Stored item 'GarageMandoorLHF' as '2017-07-17T21:36:56' in SQL database at Mon Jul 17 21:36:56 EDT 2017 in 5 ms.

**Comment by lewie** [comment]
@sciamachy, which Database do you use?

**Comment by sciamachy** [comment]
@lewie, mysql  Ver 14.14 Distrib 5.7.18, for Linux (x86_64)

**Comment by tefinger** [comment]
Same problem here. Time is always 00:00:00 in MySql Db 
MySql Server Version: 5.5.58

**Comment by 9037568** [comment]
I think the problem is actually [here](https://github.com/openhab/openhab1-addons/blob/b0320e117dd241e00ccffaf6812ac64d3f233abd/bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/db/JdbcBaseDAO.java#L325), where the persistence mistakenly relies on the default toString() implementation in the object being stored:

````
public void doStoreItemValue(Item item, ItemVO vo) {
        ...
        logger.debug("JDBC::doStoreItemValue sql={} value='{}'", sql, vo.getValue());
        Yank.execute(sql, params);
}
````

**Comment by 9037568** [comment]
This is because java.sql.Date's toString() returns only the month, day and year.

From the API doc:

````
public String toString()

Formats a date in the date escape format yyyy-mm-dd. 
````

**Comment by Moxified** [comment]
I too am seeing this issue with mariadb.  Is there any update on this?  Is a fix going to make it into OH 2.2 release?

I'm not certain why but this also seems to break my rules with 
 item.changedsince(nowminusMinutes(x))

I have rules that use jdbc persistence with changedsince that work fine for string items such as a motion sensor but these datetime items act as if they haven't changed.

**Comment by 9037568** [comment]
We've probably missed the window for getting this into 2.2.

@sciamachy, do you wish to submit a PR for your suggested change?

In the meantime, [here's a jar](https://drive.google.com/open?id=1VubglapT0geUCijWyjWxzAulWy6j4ZNM) that implements the suggested change, if anyone (@Moxified @tefinger) would care to try it out...

**Comment by Moxified** [comment]
I placed the jar file in /usr/share/openhab2/addons and restarted.  Doesn't seam to make a difference.  Is this not the proper location?  It is what I had documented from earlier testing.

I'm using apt method on debian.

**Comment by 9037568** [comment]
Not sure about that. You might need to uninstall the original version of persistence via the console.

**Comment by Moxified** [comment]
I stopped OH, removed the contents of /var/lib/openhab2/cache, placed the jar in /usr/share/openhab2/addons and started.  It installs the binding but it continues to throw errors saying JDBC:openConnection: no driver available!

I stopped it again, purged cache again, removed the jar from addons and restarted, JDBC goes back to working as it was. 

Not sure if I'm doing something wrong or if the jar is no good.

I will admit I'm new to the jdbc driver as I have been using mysql binding up to this point.  The jdbc driver is a little confusing to me as it seems that in some configuration files you list jdbc, other jdbc-mariadb.  The persist file is jdbc.persist, the default persistence has to be jdbc but the addons.cfg needs jdbc-mariadb.  Perhaps this is just my misunderstanding.  If so, I apologize.

**Comment by lewie** [comment]
Thanks for troubleshooting and PR, Chris!

@Moxified, before you place the new bundle in `/usr/share/openhab2/addons`. Please comment `persistence = jdbc-mysql` in addons.cfg. Additionally, uninstall all jdbc-* bindings over Paper-UI! If not, openHAB is trying to start a second JDBC bindings. 

All bindings placed in `/usr/share/openhab2/addons` are starting automatically. This is a third alternative beside installing bindings over Paper-UI or addons.cfg entry's.

jdbc.persist is the configuration for the JDBC Binding (there is only one). In Paper-UI you see several configurations of the JDBC binding.  jdbc-* in Papar-UI or addons.cfg refers to the driver which JDBC binding should load.

**Comment by Moxified** [comment]
TLDR: It works.

I thought I had commented it out.  It is possible I didn't wait long enough.  It seems to take a little while sometimes to load addons or remove them.

I tried again and received this error:
```
2017-12-21 16:05:50.129 [ERROR] [ence.jdbc.internal.JdbcConfiguration] - JDBC::updateConfig: could NOT load JDBC-driverClassName or JDBC-dataSourceClassName. ClassNotFoundException: 'org.mariadb.jdbc.Driver cannot be found by org.openhab.persistence.jdbc_1.11.0.201712170333'
2017-12-21 16:05:50.134 [WARN ] [ence.jdbc.internal.JdbcConfiguration] -

        !!!
        To avoid this error, place an appropriate JDBC driver file for serviceName 'mariadb' in addons directory.
        Copy missing JDBC-Driver-jar to your OpenHab/addons Folder.
        !!!
        DOWNLOAD:
        MariaDB:   version >= 1.2.0 from              http://mvnrepository.com/artifact/org.mariadb.jdbc/mariadb-java-client

2017-12-21 16:05:50.137 [WARN ] [persistence.jdbc.internal.JdbcMapper] - JDBC::openConnection: no driver available!
```

I downloaded this: http://central.maven.org/maven2/org/mariadb/jdbc/mariadb-java-client/2.2.0/mariadb-java-client-2.2.0.jar

At first it didn't seem to fix it but I walked away frustrated and it finally loaded and I was able to see time being stamped into my mariadb items now.

**Comment by lewie** [comment]
Well, then christmas is saved. ;-)
